### PR TITLE
Fix data_seed unused

### LINKED ISSUE #31818: [Potential Bug] `data_seed` in `TrainingArguments` is unused (I think)
### System Info

- `transformers` version: 4.42.3
- Platform: Linux-5.4.0-162-generic-x86_64-with-glibc2.31
- Python version: 3.12.4
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.32.1
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.3.1+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: Yes
- GPU type: NVIDIA A100-SXM4-80GB

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Run this simple script https://github.com/huggingface/trl/blob/main/examples/scripts/reward_modeling.py, sweeping `data_seed` e.g., `--data_seed=0` for one run, `--data_seed=1` for another run.

### Expected behavior

I would expect `data_seed` to make a difference. However, it does not: 

image

I opened an issue with [TRL](https://github.com/huggingface/trl/issues/1810) but after looking at `transformers` more closely, I can't actually find where `data_seed` is used. Could someone please check whether `data_seed` is actually used?

### PR DESCRIPTION
**Pull request by MekkCyber**
# What does this PR do?
This PR fixes the problem related to the data_seed being unused. It was removed during a refactor of how samplers used to work.

Fixes https://github.com/huggingface/transformers/issues/31818 https://github.com/huggingface/trl/issues/1810

You need to use this accelerate PR : https://github.com/huggingface/accelerate/pull/3130

For example in this code : 
```
from datasets import load_dataset
from transformers import AutoTokenizer, TrainingArguments, AutoModelForSequenceClassification, Trainer, set_seed
import torch

set_seed(125)

tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
model = AutoModelForSequenceClassification.from_pretrained("bert-base-uncased", num_labels=2)

dataset = load_dataset("glue", "mrpc", split="train")

def tokenize_function(example):
    return tokenizer(example["sentence1"], example["sentence2"], truncation=True)

dataset = dataset.map(tokenize_function, batched=True)
SEED = 126
DATA_SEED = 762
training_args = TrainingArguments(f"test-trainer_dataseed_{DATA_SEED}_seed_{SEED}", max_steps=400, logging_steps=40, seed=SEED, data_seed=DATA_SEED)
trainer = Trainer(
    model,
    training_args,
    train_dataset=dataset,
    tokenizer=tokenizer,
)

trainer.train()
```
When we set the seed using `set_seed`, if the data sampler's seed is chosen using `torch.random.initial_seed()`, it will remain deterministic even when the `data_seed` value changes. To address this, the PR checks whether a generator using the `data_seed` is already available. If it exists, we use the initial `data_seed` provided by the user. If not, we randomly select one. This ensures that even when `set_seed` is used, the randomness of the data sampler remains unaffected when different `data_seed` values are used.

### CODE REVIEW

**Review by vignesh1507 — CHANGES_REQUESTED** [review]
can you elaborate more about the changes you made in the code?

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -404,6 +404,18 @@ def __init__(
         self.args = args
         # Seed must be set before instantiating the model when using model
         enable_full_determinism(self.args.seed) if self.args.full_determinism else set_seed(self.args.seed)
+        
+
+        generator = torch.Generator()
+        if self.args.data_seed is not None:
+            # TODO change min version of accelerate
+            if self.args.accelerator_config.use_seedable_sampler and is_accelerate_available(min_version="0.34.0"):
+                raise RuntimeError("data_seed requires accelerate >= 0.34.0, please install the latest version of accelerate")
```
Comment (SunMarc) [inline src/transformers/trainer.py:413]: make sure to update this when we fix the issue on accelerate

**Review by SunMarc — APPROVED** [review]
LGTM as we discussed offline ! Don't forget to fix the CI with `make style`

**Review by LysandreJik — APPROVED** [review]
Let's merge if it unblocks, and Zach feel free to comment when back if it's not up to your standards

**Comment by SunMarc** [comment]
cc @MekkCyber  let's update the PR since we merged this PR : https://github.com/huggingface/accelerate/pull/3150#event-14575790028

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -404,6 +404,19 @@ def __init__(
...
+        if self.args.data_seed is not None:
+            # TODO change min version of accelerate
+            if self.args.accelerator_config.use_seedable_sampler and is_accelerate_available(min_version="0.34.0"):
+                raise RuntimeError(
+                    "data_seed requires accelerate >= 0.34.0, please install the latest version of accelerate"
+                )
+            generator.manual_seed(self.args.data_seed)
+        else:
+            generator.manual_seed(self.args.seed)
+        self.data_sampler_generator = generator
```
Comment (muellerzr) [inline src/transformers/trainer.py:418]: With the new version what should essentially happen is:

Instead we should modify `create_accelerator_and_postprocess` and modify this chunk of it https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L4860-L4866

With something like:

```python
        if is_accelerate_available("0.28.0"):
            dataloader_config = DataLoaderConfiguration(
                split_batches=accelerator_config.pop("split_batches"),
                dispatch_batches=accelerator_config.pop("dispatch_batches"),
                even_batches=accelerator_config.pop("even_batches"),
                use_seedable_sampler=accelerator_config.pop("use_seedable_sampler"),
            )
            if is_accelerate_available("1.1.0"):
                dataloader_config.data_seed = self.args.data_seed
```
And then in `TrainingArguments.__post_init__` check if a `data_seed` was passed/not the default. If so, do a check for the accelerate version and raise a `NotImplementedError()` telling users to upgrade their accelerate version.

**Review by muellerzr — COMMENTED** [review]
Some direction on where to go from here:

**Review by SunMarc — APPROVED** [review]
Nice and simple ! Let's fix the merge conflits and we are good to merge !
